src/core/thread/mle_router.cpp

```diff
@@ -3251,11 +3251,14 @@ void MleRouter::SendChildUpdateResponse(Child *                 aChild,
     }
 }
 
-void MleRouter::SendDataResponse(const Ip6::Address &aDestination,
+void MleRouter::SendDataResponse(const Message *     aMessage,
+                                 const Ip6::Address &aDestination,
                                  const uint8_t *     aTlvs,
                                  uint8_t             aTlvsLength,
                                  uint16_t            aDelay)
```

Comment:
We seems only need the `MessageMetadata` but not the full request message. 
I would suggest to pass in only the request information and use default value to avoid changes to unrelated code. 

```suggestion
void MleRouter::SendDataResponse(const Ip6::Address &aDestination,
                                 const uint8_t *     aTlvs,
                                 uint8_t             aTlvsLength,
                                 uint16_t            aDelay, 
                                 const MessageMetadata *aRequestMetadata=nullptr)
```

Thoughts? @Irving-cl @librasungirl

Reply:
Make sense to me. But `GetMetadata()` is private in `Message`. It's not exposed, so I think we'd better keep this. We can still add the `const Message *` at the end of the parameters and make it default to be `nullptr`. Thoughts? @simonlingoogle

Reply:
SGTM